File: Dynamic_Programming/1D/codeHelp/01Knapsack.cpp
#include <bits/stdc++.h>
#define endl "\n"
#pragma GCC optimize ("Ofast")
#pragma GCC target ("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx")
#pragma GCC optimize ("-ffloat-store")
using namespace std;


// Author_Name: Tipto_Ghosh
// Last Updated: 14-08-2024


typedef long long ll;
typedef long long int lli;
typedef long double ld;
typedef unsigned long long int ulli;
const int N = 1e7 + 10;
const int mod = 1e9 + 7;
const double pi = 3.14159265358979323846;
const double inf_double = numeric_limits<double>::infinity();
const int inf_int = numeric_limits<int>::max();

//------------------Solve Here----------------------------------

// solve using recursion
int solveRecursion(int currIndex , vector<int> &weight, vector<int> &value, int capacity) {
    if(currIndex == 0) {
        // do we have the capacity to take the weight yes or not
        if(capacity >= weight[currIndex]) {
           return value[currIndex]; 
        }
        else{
            // if don;t have the capacity do not take
            return 0;
        }
    }

    // take the currIndex[to take the item check , do we have capacity to take it or not]
    int include = 0;
    if(capacity >= weight[currIndex]) {
        include = value[currIndex] + solveRecursion(currIndex-1 , weight,value , capacity - weight[currIndex]);
    } 
    // don't take the currIndex
    int exclude = solveRecursion(currIndex-1 , weight , value , capacity);
  return max(include , exclude);  
}

// Using Top-Down Approch
int solveMemo(int currIndex , vector<int> &weight, vector<int> &value, int capacity , vector<vector<int>> &dp) {
  if(currIndex == 0) {
    if(capacity >= weight[currIndex]) {
        return value[currIndex];
    }
    else{
        return 0;
    }
  }
  
  // memoization part
  if(dp[currIndex][capacity] != -1) {
    return dp[currIndex][capacity];
  }

  // take the currIndex[to take the item check , do we have capacity to take it or not]
    int include = 0;
    if(capacity >= weight[currIndex]) {
        include = value[currIndex] + solveMemo(currIndex-1 , weight,value , capacity - weight[currIndex] , dp);
    } 
    // don't take the currIndex
    int exclude = solveMemo(currIndex-1 , weight , value , capacity , dp);
    // store the value
    dp[currIndex][capacity] = max(include , exclude);
  return dp[currIndex][capacity];
}

// Solve Using Tabulation
int solveTabulation(vector<int> &weight, vector<int> &value, int n, int capacity) {
    // step-1: Create the dp array
    vector<vector<int>> dp(n , vector<int> (capacity+1 , 0));

    // step-02: analyze The Base case:[for row 0 every value will value[0] if has capacity , if don't have capacity then 0]
    for(int i = 0; i <= capacity; i++) {
        if(i >= weight[0]) { // i is the curr capacity
            dp[0][i] = value[0];
        }
        else{
            dp[0][i] = 0;
        }
    }

    // solve for others
    for(int index = 1; index < n; index++) {
        for(int w = 0; w <= capacity; w++) {
            int include = 0;
            if(w >= weight[index]) {
                include = value[index] + dp[index-1][w - weight[index]];
            }
            int exclude = dp[index-1][w];

            dp[index][w] = max(include , exclude);
        }
    }
  return dp[n-1][capacity];
}
int knapsack(vector<int> weight, vector<int> value, int n, int capacity) {
  //  return solveRecursion(n-1 , weight , value , capacity);
//   vector<vector<int>> dp(n , vector<int> (capacity+1 , -1));
//   return solveMemo(n-1 , weight , value , capacity , dp);

   return solveTabulation(weight , value , n , capacity);
}



//----------------End Here--------------------------------------
int main() {
    
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    freopen("error.txt", "w", stderr);
    #endif
//---------------------------------------------------------------
    
    vector<int> weight = {1,2,4,5};
    vector<int> value = {5,4,8,6};
    int n = 4 , capacity = 5;

    cout << knapsack(weight , value , n , capacity) << endl;






   return 0;
}

File: Dynamic_Programming/Striver/Dp_On_Subsequence/01_knapsack.py
class Solution:
    def knapSack(self, capacity, val, wt):
        # n = len(val)
        # dp = [[-1] * (capacity + 1) for _ in range(n)]
        # def solve(index: int , curr_capacity: int) -> int:
        #     if index == 0:
        #         return val[0] if curr_capacity >= wt[0] else 0
        #     if dp[index][curr_capacity] != -1:
        #         return dp[index][curr_capacity]
            
        #     skip = solve(index - 1 , curr_capacity)
        #     take = 0
        #     if curr_capacity >= wt[index]:
        #         take = val[index] + solve(index - 1 , curr_capacity - wt[index])
                
        #     dp[index][curr_capacity] = max(take , skip)
        #     return dp[index][curr_capacity]
        # return solve(n-1 , capacity)
        
        # Base case
        # for every ele in the 0-th index if capacity is greater than ele then we can take it
        # n = len(val)
        # dp = [[0] * (capacity + 1) for _ in range(n)]
        # for w in range(capacity + 1): # we can take the 0-th item if capacity allows
        #     if w >= wt[0]:
        #         dp[0][w] = val[0]
        
        # for i in range(1 , n):
        #     for w in range(capacity + 1):
        #         skip = dp[i-1][w]
        #         take = 0
        #         if w >= wt[i]:
        #             take = val[i] + dp[i-1][w - wt[i]]
        #         dp[i][w] = max(take , skip)
        # return dp[n-1][capacity]        
        
        #------------Space Optimization--------------------
        n = len(val)
        prev = [0] * (capacity + 1)
        curr = [0] * (capacity + 1)     
        for w in range(capacity + 1): # we can take the 0-th item if capacity allows
            if w >= wt[0]:
                prev[w] = val[0]
        
        for i in range(1 , n):
            for w in range(capacity + 1):
                skip = prev[w]
                take = 0
                if w >= wt[i]:
                    take = val[i] + prev[w - wt[i]]
                curr[w] = max(take , skip)
            prev = curr
        return prev[capacity]        
        
sol = Solution()
val = [10 , 40 , 30 , 50]
wt = [5 , 4 , 6 , 3]

print(sol.knapSack(4 , val , wt))